i#3068: API for accessing cache model metrics

mat-kalinowski opened the pull request:
This patch contains implementation of the API for accessing separate
metrics gathered during the simulation.

Functions for accessing metrics were added for all caches in the
default configuration: L1I, L1D and LL. Also additional functions
for getting configuration options of the cache_simulator_t.

Issue: #3068

derekbruening commented on clients/drcachesim/simulator/cache_simulator.h:
```diff
@@ -60,12 +60,26 @@ class cache_simulator_t : public simulator_t {
+    int_least64_t
+    get_l1d_metric(metric_name_t metric, unsigned int core) const;
+    int_least64_t
+    get_ll_metric(metric_name_t metric) const;
```
In our usage we routinely have a 3-level cache with a per-core L2 in between L1{I,D} and LLC (created with a config file).  Possibilities:

+ Add a get_l2_metric().  To take in a core index, it would have to only support direct parents of L1* and so would not support arbitrary cache layouts.
+ Export an iterator over all caches (split into l1, llc, and "other"==l2 separate iterators?) which returns some kind of key (the cache name?) to then pass back to query a metric.
+ Take in the cache by name (the name from the config file) -- less usable by itself without an iterator.

mat-kalinowski replied:
Thanks for the suggestions! Sorry for the delay, I had few days off. I like the flexibility of accessing stats in arbitrary cache model and also having really simple interface. I thought about having common function for all caches, something like `int_least64_t get_cache_metric(unsigned level, unsigned core, metric_t metric)`
Does separation for instruction and data caches makes sense only for L1 cache in any model ?

derekbruening replied:
I think the current code only supports i vs d at L1 but if it's easy not to assume that here that might help future support.

> something like `int_least64_t get_cache_metric(unsigned level, unsigned core, metric_t metric)`

Would this support asymmetric layouts?  I guess it would if you always walk from the core up `level` levels.

mat-kalinowski replied:
I tried to make this code as flexible as possible. Current approach should work with split cache of the higher level in simple models, when given higher level split cache is an ancestor of L1 (for example L1i -> L2i).

mat-kalinowski replied:
Can you provide example of some simple asymmetric layout ? My current approach is to walk up from L1 cache and I think it should work properly.

derekbruening replied:
> Can you provide example of some simple asymmetric layout ? My current approach is to walk up from L1 cache and I think it should work properly.

Never used or seen anything too weird with dr$sim, was just speculating about possible corner cases.  Such a walk should work with any layout though it seems.

derekbruening commented on clients/drcachesim/simulator/cache_simulator.h:
```diff
@@ -60,12 +60,26 @@ class cache_simulator_t : public simulator_t {
     remaining_sim_refs() const;
 
+    unsigned int
+    get_core_num() const;
```
Since cache_simulator_knobs_t is already public -- should the knobs be returned?  IIRC a config file setup does set the key knobs fields?

mat-kalinowski replied:
Yes, knobs fields are set by config file setup. I have added suitable method.

derekbruening commented on clients/drcachesim/simulator/caching_device_stats.h:
```diff
@@ -115,6 +141,10 @@ class caching_device_stats_t {
+    // References to the properties with statistics are held in the map with the
+    // statistic name as the key. Sample map element: {HITS, num_hits_}
+    std::map<metric_name_t, int_least64_t &> stats_map_;
+
```
Please add some kind of test.  Maybe in drcachesim_unit_tests or a new similar test?  Xref #4842.

mat-kalinowski replied:
I have added unit test in the `drcachesim_unit_tests`. I couldn't test `Invalidates` metric properly because of the bug/issue that I described here: https://github.com/DynamoRIO/dynamorio/issues/5031

derekbruening replied:
OK, best to put a `XXX i#5031: ` or `TODO i#5031: ` comment in the test about adding that metric.

derekbruening reviewed (commented):
Thank you for the contribution.  I have a few suggestions.

AssadHashmi commented:
add to whitelist

derekbruening commented:
Looks like #4954 was hit on A64 Jenkins -- maybe should be added to the known-flaky list.

derekbruening commented:
run arm tests

mat-kalinowski commented:
Right now I am handling errors like this: 
```
if (parent == NULL) {
    ERRMSG("Wrong cache level: %u.\n", core);
    return 0;
}
```
I thought also about returning -1 without printing error message. Which approach you think is better? Also do you have some comments about newest commit with changes ?

derekbruening commented:
> Right now I am handling errors like this:
> 
> ```
> if (parent == NULL) {
>     ERRMSG("Wrong cache level: %u.\n", core);
>     return 0;
> }
> ```
> 
> I thought also about returning -1 without printing error message. Which approach you think is better? 

For get_cache_metric() I would think of it as a library routine likely to be used by third-party code and as such I'd return an error rather than printing an error.

> Also do you have some comments about newest commit with changes ?

Oh, was waiting for a (re-)review request (https://dynamorio.org/page_code_reviews.html#autotoc_md114 "When the requested changes have been pushed, you can ask for a re-review from the reviewers. This can be done by clicking the re-review button next to the reviewers' name on the right sidebar on the pull request page.")

derekbruening commented on clients/drcachesim/simulator/cache_simulator.h:
```diff
@@ -43,6 +43,8 @@
 #include "cache.h"
 #include "snoop_filter.h"
 
+enum cache_split_t { DATA, INSTRUCTION };
```
Safer to put inside the class and not in the global namespace (xref #4343).  Or see other suggestions for the enum below.

derekbruening commented on clients/drcachesim/simulator/cache_simulator.h:
```diff
@@ -60,12 +62,19 @@ class cache_simulator_t : public simulator_t {
     print_results() override;
 
+    int_least64_t
+    get_cache_metric(unsigned level, metric_name_t metric, unsigned core = 0,
```
I would re-order the parameters with the metric first and grouping the 3 identifiers of which cache (level, core, split) together.

derekbruening commented on clients/drcachesim/simulator/caching_device_stats.h:
```diff
@@ -49,6 +50,20 @@ enum invalidation_type_t {
     INVALIDATION_COHERENCE,
 };
 
+enum metric_name_t {
```
I know the existing code has global scope enums but we're trying to avoid that going forward (xref #4343), and here we have very simple names that could conflict.  We do require C++11 so this could be a scoped enum (`typedef enum class metric_name_t`) where it would require the prefix `metric_name_t::HITS`; or at least use a prefix like the existing enums so `METRIC_HITS`; or something, to avoid global symbols with very simple names like `HITS` which could conflict (this code is statically linked into some very large code bases).

derekbruening commented on clients/drcachesim/simulator/cache_simulator.cpp:
```diff
@@ -594,6 +594,48 @@ cache_simulator_t::print_results()
+
+    if (core >= knobs_.num_cores) {
+        ERRMSG("Wrong core number: %u.\n", core);
+        return 0;
```
Xref the top-level thread: I would vote for returning an error.  Could go so far as error enum types; or just return -1 (document that's the error return code) either w/ or w/o the print (could see the print useful if there's no errno; could see it annoying linked into a large system that maybe doesn't know the cache layout and is passing all possible levels and using return code to find the bounds of the layout -- there an errno would be nice).

Ditto below.

derekbruening reviewed (commented):
Thank you for adding the test.  Just a few comments mostly about names and then should be good to go.

mat-kalinowski commented:
I've changed enum types to the `enum class` to prevent polluting global namespaces. I've implemented `stats_error_t` as a plain enum and added `STATS_` prefix because I needed to return it as an int from the `get_cache_metric` function.

derekbruening commented on clients/drcachesim/simulator/cache_simulator.h:
```diff
@@ -63,8 +75,8 @@ class cache_simulator_t : public simulator_t {
     int_least64_t
-    get_cache_metric(unsigned level, metric_name_t metric, unsigned core = 0,
-                     cache_split_t split = DATA) const;
+    get_cache_metric(metric_name_t metric, unsigned level, unsigned core = 0,
```
Please comment that a negative return value is an error and is of type stats_error_t

derekbruening commented on clients/drcachesim/simulator/cache_simulator.h:
```diff
@@ -42,8 +42,20 @@
+enum class cache_split_t { DATA, INSTRUCTION };
+
+// Error codes returned when passing wrong parameters to the function for
+// accessing cache statistics.
```
Use the name `get_cache_metric` rather than a description so a reader knows precisely which one

derekbruening reviewed (approved):
I just have some comment suggestions, otherwise looks good.  Thank you for the contribution.

mat-kalinowski commented:
run arm tests

derekbruening commented:
Thanks again for the contribution.

One nit: For future merges please remove the duplicate title line (xref https://dynamorio.org/page_code_reviews.html#autotoc_md117):

```
commit 9574fca502ee03661dfd788f8744a8881f19eb63

    i#3068: API for accessing cache model metrics (#5020)
    
    i#3068: API for accessing cache model metrics
    
    This patch contains implementation of the API for accessing separate
    metrics gathered during the simulation.
    
    Functions for accessing metrics were added for all caches in the
    default configuration: L1I, L1D and LL. Also additional functions
    for getting configuration options of the cache_simulator_t.
    
    Issue: #3068
```
